### PR TITLE
Changes in doc concerning the use of CMIP6-like custom tables

### LINKED ISSUE #2364: Using .json custom table
**Is your feature request related to a problem? Please describe.**
Here at CNRM we have a few "personal" CMOR tables written in json format (example attached). I tried to use them as custom tables in my recipes playing with option like `cmor_path` in the `config-developer.yml` file but unsuccessfully. I would like to know if it is possible or easy to implement the possibility to use such files as custom tables in esmvalcore (I understood from the documentation that the CMIP5 .dat format was in use for custom tables). 

**Would you be able to help out?**
I have some time to help out but I have basic/intermediate knowledge of python and a beginner with esmvaltool. 

[CNRM_HOMP39Amon.json](https://github.com/ESMValGroup/ESMValCore/files/14512184/CNRM_HOMP39Amon.json)

### PR DESCRIPTION
**Pull request by MartinCussacMF**
## Description

Modifications to the 'configure' documentation, to better reflect the possibility to use both CMIP5-like and CMIP6-format for custom CMOR tables. Added an example of a `.json` CMIP-like custom table. 

Closes #2364

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [ ] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: doc/quickstart/configure.rst [hunk 1]
```diff
@@ -722,9 +722,9 @@ or as absolute path.
 Other options given for this special table will be ignored.
 
 Custom tables in this directory need to follow the naming convention
-``CMOR_{short_name}.dat`` and need to be given in CMIP5 format.
+``CMOR_{short_name}.dat`` or ``CMOR_{short_name}.dat``, and need to be given in the corresponding CMIP5 or CMIP6 format.
```
Comment (valeriupredoi) [inline doc/quickstart/configure.rst:725]:
```suggestion
``CMOR_{short_name}.dat`` or ``CMOR_{short_name}.json``, and need to be given in the corresponding CMIP5 or CMIP6 format.
```

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
hi @MartinCussacMF many thanks for your PR! Just a small wiggle I saw in the text, @schlunma or @bouweandela do we have implementation for JSON-loading of the CMOR custom variable tables - I see it for obs4mips though not for CMIP vars :beer:

**Comment by MartinCussacMF** [comment]
Hi @valeriupredoi ! Good catch for the mistake, I accepted the sugested changes.

**Comment by bouweandela** [comment]
>  @schlunma or @bouweandela do we have implementation for JSON-loading of the CMOR custom variable tables

I'm not very up to date with this part of the code. It looks like the features were introduced by @schlunma in #1625 and #2279. @schlunma Would you have time to review this?

**Comment by schlunma** [comment]
In these PRs I only made the location of these tables customizable. This did NOT change the code that actually reads the tables.

To my knowledge the custom tables need to be in CMIP5 format (the custom table class [inherits](https://github.com/ESMValGroup/ESMValCore/blob/3a0b49f84b156c876e6785af0293feb5cb5e8445/esmvalcore/cmor/table.py#L986) from the CMIP5 class). Also, it [looks](https://github.com/ESMValGroup/ESMValCore/blob/3a0b49f84b156c876e6785af0293feb5cb5e8445/esmvalcore/cmor/table.py#L1032) like only files in the form `*.dat` are considered, so I am really surprised that this works with `*.json` files in CMIP6 format.

@MartinCussacMF are you 100% sure that works with `*.json` files? Please keep in mind that the ESMValCore might also look in other tables then the one specified by the `mip` attribute if a project has `cmor_strict: false` in the config-developer file.

**Comment by MartinCussacMF** [comment]
@schlunma I'm pretty sure that works with `*.json` files, and I did so following the advice from @ehogan (see issue #2364).
@valeriupredoi I checked some of the boxes I thought were relevant, don't hesitate to correct me if I'm wrong about some.

**Comment by schlunma** [comment]
All right, I will have a look at that tomorrow.

**Comment by valeriupredoi** [comment]
cheers fellas! Boxes look good @MartinCussacMF :+1: Now, the less important thing is to see if JSONs are actually loadable :grin:

**Comment by ehogan** [comment]
We use custom tables in CMIP6 format; `cmor_type: 'CMIP6'` needs specifying in the `config-developer.yml` file 🥳

**Comment by schlunma** [comment]
@ehogan could you maybe send your corresponding config-developer.yml file? I am not entirely sure if I understand you correctly.

**Comment by ehogan** [comment]
> @ehogan could you maybe send your corresponding config-developer.yml file? I am not entirely sure if I understand you correctly.

Sure thing! Here's the diff that we use (we created an `ESMVal` project to use for CMEW):

```diff
24a25,26
> # custom:
> #   cmor_path: ~/my/own/custom_tables
27,42d28
< custom:
<   cmor_path: /mip_tables/GCModelDev/0.0.9
< 
< ESMVal:
<   cmor_strict: true
<   input_dir:
<     default: '/'
<     BADC: '{activity}/{institute}/{dataset}/{exp}/{ensemble}/{mip}/{short_name}/{grid}/{version}'
<     DKRZ: '{activity}/{institute}/{dataset}/{exp}/{ensemble}/{mip}/{short_name}/{grid}/{version}'
<     ESGF: '{project}/{activity}/{institute}/{dataset}/{exp}/{ensemble}/{mip}/{short_name}/{grid}/{version}'
<     ETHZ: '{exp}/{mip}/{short_name}/{dataset}/{ensemble}/{grid}/'
<     SYNDA: '{activity}/{institute}/{dataset}/{exp}/{ensemble}/{mip}/{short_name}/{grid}/{version}'
<   input_file: '{short_name}_{mip}_{dataset}_{exp}_{ensemble}_{grid}*.nc'
<   output_file: '{project}_{dataset}_{mip}_{exp}_{ensemble}_{short_name}_{grid}'
<   cmor_type: 'CMIP6'
<   cmor_default_table_prefix: 'GCModelDev_'
```

It works for us! 🤪

**Comment by schlunma** [comment]
Hmm...I'm even more confused now 😆

Sorry for all the questions, but would you also be able to share an example recipe, and one of the table files? These table files are then only stored in `/mip_tables/GCModelDev/0.0.9`? There is no `esmvalcore/cmor/tables/esmval`? I really don't get how the `ESMVal` project can access the `cmor_path` defined in the `custom` table, especially since it's defined with `cmor_strict: true`...

**Comment by ehogan** [comment]
> Hmm...I'm even more confused now 😆

Oh no, sorry! 😝 

> Sorry for all the questions, but would you also be able to share an example recipe, and one of the table files? These table files are then only stored in `/mip_tables/GCModelDev/0.0.9`? 

Ah, yes, we did update the `project` in one of the `datasets` in the recipe (we're using the radiation budget recipe for the CMEW prototype) from `CMIP6` to `ESMVal` (to match with what was defined in the `config-developer.yml` file). The table files are a direct copy and duplicate of all the CMIP6 MIP tables (including the CVs), with some extra variables added to some of the tables (that the MO are using for model development purposes, but that CMEW isn't using right now), so if you made a copy of all the CMIP6 tables into a new location and added that to the `cmor_path` then that should reflect our setup 👍 

> There is no `esmvalcore/cmor/tables/esmval`?

Correct 👍 

> I really don't get how the `ESMVal` project can access the `cmor_path` defined in the `custom` table, especially since it's defined with `cmor_strict: true`...

That I don't know! 🤔 I assumed `cmor_strict` referred to the CMOR checks, which would be just as strict, since we are using exact copies of the CMIP6 MIP tables 🤷

**Comment by ehogan** [comment]
I also meant to add that we could demo CMEW at the upcoming workshop, if that would be helpful? 😊

**Comment by schlunma** [comment]
Thanks @ehogan, that already clears things up! 😄 

One more (hopefully last) question: since you mentioned that you are using exact copies of the CMIP6 tables with some additional variables, but you are not using those additional variables right now, are you sure that the tables in `/mip_tables/GCModelDev/0.0.9` are used? With the way you set up the ESMVal project (`cmor_type: CMIP6`), the CMIP6 tables would be read by default (i.e., the ones in `esmvalcore/cmor/tables/cmip6`). You could for example change an entry from a variable you are using and check if that is reflected in your output.

From the [code](https://github.com/ESMValGroup/ESMValCore/blob/d128753fd6f6d2a10ba9b2b0c500b369fff1dd95/esmvalcore/cmor/table.py#L189), I think you need to specify `cmor_path` for the ESMVal project, i.e.,

```yaml
ESMVal:
  cmor_path: /mip_tables/GCModelDev/0.0.9
  cmor_strict: true
  ...
```

> I assumed cmor_strict referred to the CMOR checks, which would be just as strict, since we are using exact copies of the CMIP6 MIP tables 🤷

No, `cmor_strict` only controls if other variables or the custom tables may be searched. I also find the name slightly misleading 😬 

From the [doc](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#project-cmor-table-configuration): "cmor_strict: if this is set to false, the CMOR table will be extended with variables from the [Custom CMOR tables](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#custom-cmor-tables) (by default loaded from the esmvalcore/cmor/tables/custom directory) and it is possible to use variables with a mip which is different from the MIP table in which they are defined. Note that this option is always enabled for [derived](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html#variable-derivation) variables."

**Comment by ehogan** [comment]
> Thanks @ehogan, that already clears things up! 😄

Yay! 🥳 

> One more (hopefully last) question: since you mentioned that you are using exact copies of the CMIP6 tables with some additional variables, but you are not using those additional variables right now, are you sure that the tables in `/mip_tables/GCModelDev/0.0.9` are used? With the way you set up the ESMVal project (`cmor_type: CMIP6`), the CMIP6 tables would be read by default (i.e., the ones in `esmvalcore/cmor/tables/cmip6`). You could for example change an entry from a variable you are using and check if that is reflected in your output.

I believe they are, since the CVs in `GCModelDev` contain changes that are reflected in our outputs; compare some of the global attributes in the output that uses the `GCModelDev` MIP tables:
```
% ncdump -h single_value_radiation_budget/rsdt/ESMVal_UKESM1-0-LL_Amon_amip_r1i1p1f1_rsdt_gn_1993-1993.nc
netcdf ESMVal_UKESM1-0-LL_Amon_amip_r1i1p1f1_rsdt_gn_1993-1993 {
[...]
// global attributes:
		:activity_id = "ESMVal" ;
		:cmor_version = "3.7.3" ;
		:cv_version = "GCModelDev v0.0.9" ;
		:data_specs_version = "GCModelDev v0.0.9" ;
		:further_info_url = "none" ;
		:license = "GCModelDev model data is licensed under the Open Government License v3 (https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/)" ;
		:mip_era = "GCModelDev" ;
		:table_info = "Creation Date:(28 May 2020) MD5:136a5527cb291b78498f710725f53511" ;
		:title = "UKESM1-0-LL output prepared for GCModelDev" ;
```
with the global attributes in the output that uses the `CMIP6` MIP tables:
```
% ncdump -h single_value_radiation_budget/rsdt/CMIP6_HadGEM3-GC31-LL_Amon_historical_r1i1p1f3_rsdt_gn_1993-1993.nc
netcdf CMIP6_HadGEM3-GC31-LL_Amon_historical_r1i1p1f3_rsdt_gn_1993-1993 {
[...]
// global attributes:
		:activity_id = "CMIP" ;
		:cmor_version = "3.4.0" ;
		:cv_version = "6.2.20.1" ;
		:data_specs_version = "01.00.29" ;
		:further_info_url = "https://furtherinfo.es-doc.org/CMIP6.MOHC.HadGEM3-GC31-LL.historical.none.r1i1p1f3" ;
		:license = "CMIP6 model data produced by the Met Office Hadley Centre is licensed under a Creative Commons Attribution-ShareAlike 4.0 International License (https://creativecommons.org/licenses). Consult https://pcmdi.llnl.gov/CMIP6/TermsOfUse for terms of use governing CMIP6 output, including citation requirements and proper acknowledgment. Further information about this data, including some limitations, can be found via the further_info_url (recorded as a global attribute in this file) and at https://ukesm.ac.uk/cmip6. The data producers and data providers make no warranty, either express or implied, including, but not limited to, warranties of merchantability and fitness for a particular purpose. All liabilities arising from the supply of the information (including any liability arising in negligence) are excluded to the fullest extent permitted by law." ;
		:mip_era = "CMIP6" ;
		:table_info = "Creation Date:(13 December 2018) MD5:2b12b5db6db112aa8b8b0d6c1645b121" ;
		:title = "HadGEM3-GC31-LL output prepared for CMIP6" ;
```
 
> > I assumed cmor_strict referred to the CMOR checks, which would be just as strict, since we are using exact copies of the CMIP6 MIP tables 🤷
> 
> No, `cmor_strict` only controls if other variables or the custom tables may be searched. I also find the name slightly misleading 😬

Ah, so maybe the variable tables aren't being used, but the CVs are? I will check this ...

> From the [code](https://github.com/ESMValGroup/ESMValCore/blob/d128753fd6f6d2a10ba9b2b0c500b369fff1dd95/esmvalcore/cmor/table.py#L189), I think you need to specify `cmor_path` for the ESMVal project, i.e.,
> 
> ```yaml
> ESMVal:
>   cmor_path: /mip_tables/GCModelDev/0.0.9
>   cmor_strict: true
>   ...
> ```

Hopefully this is the solution if I find the variable tables aren't being used! 😊

**Comment by MartinCussacMF** [comment]
Hello @schlunma !
Sorry I haven't really been available last few days. Are you still interested in me uploading the configuration in wich I'm able to perform a recipe with custom CMIP6 like tables?

**Comment by ehogan** [comment]
> Sorry, one more comment/question from me 😄
> 
> I think these attributes are not written by ESMValTool.

You are absolutely correct, apologies for my slowness! 😴 

I have proven to myself that my custom MIP tables are not being used by ESMValTool. When I move `cmor_path` into my `ESMVal` project in the `config-developer.yml` file I get the error:

```
Unable to load CMOR table (project) 'ESMVal' for variable 'hfls' with mip 'Amon'
```

which I what you are expecting (I'm guessing?) 🤔 I have tried to play around with the naming (since our MIP tables have the prefix `GCModelDev`) but I haven't gotten it working yet.

> Are you still interested in me uploading the configuration in wich I'm able to perform a recipe with custom CMIP6 like tables?

@MartinCussacMF, I am now also interested! 🤪

**Comment by valeriupredoi** [comment]
well this escalated quickly :rofl:

**Comment by MartinCussacMF** [comment]
Well here is the awaited recipe/config I'm using atm:
-  the recipe: [recipe_zonal_sig_emich4.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920052/recipe_zonal_sig_emich4.yml.txt)
- config files: [config-user.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920061/config-user.yml.txt) & [config-developer.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920063/config-developer.yml.txt)
- custom table and coordinates: [CMOR_HOMP39Amon.json](https://github.com/ESMValGroup/ESMValCore/files/14920094/CMOR_HOMP39Amon.json) & [CMOR_coordinate.json](https://github.com/ESMValGroup/ESMValCore/files/14920097/CMOR_coordinate.json)
- and finally the log: [main_log.txt](https://github.com/ESMValGroup/ESMValCore/files/14920104/main_log.txt)

Let me know if you have questions/want some more info.

**Comment by ehogan** [comment]
I figured it out! 🥳 There were two issues:

1) I did need to move `cmor_path` into the `ESMVal` project in the `config-developer.yml` file, as @schlunma suggested in https://github.com/ESMValGroup/ESMValCore/pull/2368#issuecomment-2044556436; thank you! 😁 Other than that, the differences to my `config-developer.yml` file are still as in https://github.com/ESMValGroup/ESMValCore/pull/2368#issuecomment-2043038135 (is it possible that `cmor_strict: true` works in this case because `cmor_path` is specified directly within the project and provides a complete copy of all MIP tables?)
1) ESMValTool requires that the MIP tables are stored in a `Tables` directory within the path provided by `cmor_path`; would it be possible to relax this / make it configurable?

Evidence:
```
% ncdump -h recipe_radiation_budget_20240409_181752/preproc/single_value_radiation_budget/rsdt/ESMVal_UKESM1-0-LL_Amon_esmval-dummy-a_r1i1p1f1_rsdt_gn_1993-1993.nc
netcdf ESMVal_UKESM1-0-LL_Amon_esmval-dummy-a_r1i1p1f1_rsdt_gn_1993-1993 {
dimensions:
	bnds = 2 ;
variables:
	float rsdt ;
		rsdt:_FillValue = 1.e+20f ;
		rsdt:standard_name = "toa_incoming_shortwave_flux" ;
		rsdt:long_name = "Will my edit show in the output netCDF file? I hope so!" ;
		rsdt:units = "W m-2" ;
		rsdt:cell_methods = "area: time: mean longitude: latitude: mean time: mean" ;
		rsdt:coordinates = "lat lon time" ;
```
🤪🥳

**Comment by schlunma** [comment]
> Well here is the awaited recipe/config I'm using atm:
> 
>     * the recipe: [recipe_zonal_sig_emich4.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920052/recipe_zonal_sig_emich4.yml.txt)
> 
>     * config files: [config-user.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920061/config-user.yml.txt) & [config-developer.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920063/config-developer.yml.txt)
> 
>     * custom table and coordinates: [CMOR_HOMP39Amon.json](https://github.com/ESMValGroup/ESMValCore/files/14920094/CMOR_HOMP39Amon.json) & [CMOR_coordinate.json](https://github.com/ESMValGroup/ESMValCore/files/14920097/CMOR_coordinate.json)
> 
>     * and finally the log: [main_log.txt](https://github.com/ESMValGroup/ESMValCore/files/14920104/main_log.txt)
> 
> 
> Let me know if you have questions/want some more info.

Thanks for that @MartinCussacMF!

Sorry for being so stubborn, but with this evidence I am pretty sure that you custom tables are also **NOT** read (similar to @ehogan's case).

Here's what's happening (I think):
- You defined your custom **project** CNRM with `cmor_type: 'CMIP6'` (note that this is very different to what we call custom **tables**; custom tables are simply another directory which can be used by all derived variables and any other project defined with `cmor_strict: false`). Since you did not specify a `cmor_path` for this, the tables located at `esmvalcore/cmor/tables/cmip6/Tables` are read by default.
- ESMValTool will now look for a table `esmvalcore/cmor/tables/cmip6/Tables/CMIP6_HOMP39Amon` (your `mip` defined in the recipe), which it does not find. However, since project CNRM is defined with `cmor_strict: false`, ESMValTool is [allowed to search other mips](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#project-cmor-table-configuration). One appearance of the variable `ch4` is for example in the [CMIP6 Amon](https://github.com/ESMValGroup/ESMValCore/blob/d128753fd6f6d2a10ba9b2b0c500b369fff1dd95/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_Amon.json#L107-L124) table, which will now be used.

As I've already mentioned to @ehogan, it's easy to verify this: could you please alter the `long_name` of ch4 in your `CMOR_HOMP39Amon.json` table and check if that is reflected in the output of the tool (the files located in the `preproc` directory). Alternatively, you could remove the entire ch4 block in your custom tables or use `cmor_strict: true` for your CNRM project. If the run fails, your tables have been read. If the run still works, your tables are not read.

**Comment by schlunma** [comment]
> I figured it out! 🥳 There were two issues:

That's great to hear @ehogan!

> is it possible that `cmor_strict: true` works in this case because `cmor_path` is specified directly within the project and provides a complete copy of all MIP tables?

The main reason it works is that you have a complete copy of the MIP tables. Without `cmor_path`, the default CMIP6 tables would be read, so it would still work as long as you only use CMIP6 variables (you involuntary tested this earlier 😁). `cmor_strict: true` means that the tool will only check the MIP you defined in the recipe, e.g., for `mip: Amon`, only the table `..._Amon.json` is searched. If a variable is not there, an error will be raised. With `cmor_strict: false`, this is relaxed, and also other MIP tables **of the same project** (e.g., `..._Lmon.json`) and the custom tables are searched, but **never** tables from another project.

> 2. ESMValTool requires that the MIP tables are stored in a `Tables` directory within the path provided by `cmor_path`; would it be possible to relax this / make it configurable?

I think it's certainly possible, but let's discuss that in another issue 👍

**Comment by MartinCussacMF** [comment]
> > Well here is the awaited recipe/config I'm using atm:
> > ```
> > * the recipe: [recipe_zonal_sig_emich4.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920052/recipe_zonal_sig_emich4.yml.txt)
> > 
> > * config files: [config-user.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920061/config-user.yml.txt) & [config-developer.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14920063/config-developer.yml.txt)
> > 
> > * custom table and coordinates: [CMOR_HOMP39Amon.json](https://github.com/ESMValGroup/ESMValCore/files/14920094/CMOR_HOMP39Amon.json) & [CMOR_coordinate.json](https://github.com/ESMValGroup/ESMValCore/files/14920097/CMOR_coordinate.json)
> > 
> > * and finally the log: [main_log.txt](https://github.com/ESMValGroup/ESMValCore/files/14920104/main_log.txt)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Let me know if you have questions/want some more info.
> 
> Thanks for that @MartinCussacMF!
> 
> Sorry for being so stubborn, but with this evidence I am pretty sure that your custom tables are also **NOT** read (similar to @ehogan's case).

No worries, I'm rather new to the code so there are things I don't get right away. 

> 
> Here's what's happening (I think):
> 
>     * You defined your custom **project** CNRM with `cmor_type: 'CMIP6'` (note that this is very different to what we call custom **tables**; custom tables are simply another directory which can be used by all derived variables and any other project defined with `cmor_strict: false`). Since you did not specify a `cmor_path` for this, the tables located at `esmvalcore/cmor/tables/cmip6/Tables` are read by default.
> 
>     * ESMValTool will now look for a table `esmvalcore/cmor/tables/cmip6/Tables/CMIP6_HOMP39Amon` (your `mip` defined in the recipe), which it does not find. However, since project CNRM is defined with `cmor_strict: false`, ESMValTool is [allowed to search other mips](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#project-cmor-table-configuration). One appearance of the variable `ch4` is for example in the [CMIP6 Amon](https://github.com/ESMValGroup/ESMValCore/blob/d128753fd6f6d2a10ba9b2b0c500b369fff1dd95/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_Amon.json#L107-L124) table, which will now be used.
> 
> 
> As I've already mentioned to @ehogan, it's easy to verify this: could you please alter the `long_name` of ch4 in your `CMOR_HOMP39Amon.json` table and check if that is reflected in the output of the tool (the files located in the `preproc` directory). Alternatively, you could remove the entire ch4 block in your custom tables or use `cmor_strict: true` for your CNRM project. If the run fails, your tables have been read. If the run still works, your tables are not read.

Indeed, I can't see the modified long name in the preproc directory even though I changed it in `CMOR_HOMP39Amon.json`, and my run crashes if I use  `cmor_strict: true` for my CNRM project. 
However there are still hints that esmvaltool is using the right HOMP39Amon files (sorry, I just want to undestand what esmvaltool is doing...):
- ncdump return the following, where you can clearly see that the plev has a lenght of 39 which corresponds to the HOMP39Amon grid and not the Amon one, which only has 19 vertical levels.
``` ncdump -h CNRM_CNRM-ESM2-1_HOMP39Amon_AESM-amip_ARPV7-1.03c_piclim_racmobus_emich4_2.5TgN_r1i1p1f2_ch4_gr_1850-1879.nc
netcdf CNRM_CNRM-ESM2-1_HOMP39Amon_AESM-amip_ARPV7-1.03c_piclim_racmobus_emich4_2.5TgN_r1i1p1f2_ch4_gr_1850-1879 {
dimensions:
	time = 360 ;
	plev = 39 ;
	lat = 90 ;
	bnds = 2 ;
variables:
	float ch4(time, plev, lat) ;
		ch4:_FillValue = 1.e+20f ;
		ch4:standard_name = "mole_fraction_of_methane_in_air" ;
		ch4:long_name = "Mole Fraction of CH4" ;
		ch4:units = "mol mol-1" ;
		ch4:cell_methods = "area: time: mean longitude: mean" ;
		ch4:coordinates = "lon" ;
```
- I have been able to plot some variables such as no2 for which I only have output on the HOMP39Amon table. 

Is it possible that esmvaltool uses the Amon table for the variable attributes but still uses the HOMP39Amon files found in my experiment directory ? 
Thanks for your patience.

**Comment by schlunma** [comment]
Thanks for confirming 👍 

> * ncdump return the following, where you can clearly see that the plev has a lenght of 39 which corresponds to the HOMP39Amon grid and not the Amon one, which only has 19 vertical levels.

That's because you are using the following preprocessor:

```yaml
pp27:
    extract_levels:
      levels: {cmor_table: CMIP6, coordinate: plev39}
      coordinate: air_pressure
      scheme: linear
```

This will extract the levels defined by the `plev39` coordinate of CMIP6, which has 39 levels. If you don't use this preprocessor, ESMValTool will simply use the input levels without modifying them. If they don't match the number of desired levels in the table, you will only get a warning (no error).

> * I have been able to plot some variables such as no2 for which I only have output on the HOMP39Amon table.

As mentioned, since `cmor_strict: false`, ESMValTool will search every other mip in the CMIP6 project, and it happens to find `no2` in the [AERmon table](https://github.com/ESMValGroup/ESMValCore/blob/e2c68e71e51d9bb70f614de525b14f57310128cc/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_AERmon.json#L1331-L1348).

> Is it possible that esmvaltool uses the Amon table for the variable attributes but still uses the HOMP39Amon files found in my experiment directory ? Thanks for your patience.

Nope, I am 100% sure now that the HOMP39Amon are not used in this case. If you want to use them, please have a look at [Emma's comment](https://github.com/ESMValGroup/ESMValCore/pull/2368#issuecomment-2045835772). It basically boils down to adding a `cmor_path` to your CNRM project.

I am closing this now since these changes are not necessary anymore (custom tables need to be in CMIP5 format). Feel free to reopen if you think that other changes in the documentation are necessary to clear this up!

**Comment by MartinCussacMF** [comment]
Thanks for the explanation. I'm not sure I get everything since i still am ploting some fields that I only output through the HOMP39Amon table and no other one like Amon or AERmon. 
Anyway, is it worth keeping issue #2364 open ? Loading `.json` custom table would be nice since they are the CMIP6 standard.

**Comment by schlunma** [comment]
Can you post a list of those variables here?

> Anyway, is it worth keeping issue #2364 open ? Loading `.json` custom table would be nice since they are the CMIP6 standard.

Yes, definitely!

**Comment by MartinCussacMF** [comment]
For instance, for the following recipe: [recipe_zonal_sig_no2.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/14934589/recipe_zonal_sig_no2.yml.txt)
I plot some zonal means for `no2`, and in my experiment directory I only have the following file for this variable (this is the result of `ls no2*`):
```
no2_HOMP39Amon_CNRM-ESM2-1_AESM-amip_r1i1p1f2_gr_197901-201412.nc
```
The behaviour is the same for the `co` variable for instance, the recipe is successful and I get the expected plots.

**Comment by schlunma** [comment]
You have to differentiate between finding the data (i.e., the files) and finding an appropriate variable in the CMOR tables:

- In your recipe, you have `mip: HOMP39Amon`. Thus, the tool will find data located at `no2_HOMP39Amon_CNRM-ESM2-1_AESM-amip_r1i1p1f2_gr_197901-201412.nc`.
- When looking for CMOR tables, the tool will first look for a table `CMIP6_HOMP39Amon.json` in the default CMIP6 tables, which it does NOT find. However, since `cmor_strict: false`, it will also look into other tables. It will then find the variable in the `AERmon` table, which it will use. Note that this only affects the variables; for ESMValTool, the `mip` facet will stay `HOMP39Amon`.

**Comment by MartinCussacMF** [comment]
Ok I get it now, thanks again for your time @schlunma !

**Pull request closed without merging**